node/network/statement-distribution/src/lib.rs

```diff
@@ -1064,53 +1032,65 @@ async fn handle_incoming_message_and_circulate<'a>(
 	ctx: &mut impl SubsystemContext,
 	req_sender: &mpsc::Sender<RequesterMessage>,
 	metrics: &Metrics,
-) -> Option<(&'a mut ActiveHeadData, SignedFullStatement)> {
+) -> Option<SignedFullStatement> {
 
-	let metadata = message.get_metadata();
+	let fingerprint = message.get_fingerprint();
+	let candidate_hash = *fingerprint.0.candidate_hash();
 
-	let active_head = match active_heads.get_mut(&metadata.relay_parent) {
-		Some(h) => h,
-		None => {
-			tracing::debug!(
-				target: LOG_TARGET,
-				requested_relay_parent = %metadata.relay_parent,
-				"our view out-of-sync with active heads; head not found",
-			);
-			report_peer(ctx, peer, COST_UNEXPECTED_STATEMENT).await;
-			return None
-		}
-	};
+	// Immediately return any Seconded statement:
+	let message =
+		if let protocol_v1::StatementDistributionMessage::Statement(h, s) = message {
+			if let Statement::Seconded(_) = s.payload() {
+				return Some(s)
+			}
+			protocol_v1::StatementDistributionMessage::Statement(h, s)
+		} else {
+			message
+		};
 
-	match message {
-		protocol_v1::StatementDistributionMessage::Statement(_, statement) =>
-			return Some((active_head, statement)),
-
-		protocol_v1::StatementDistributionMessage::LargeStatement(metadata) => {
-
-			let candidate_hash = metadata.candidate_hash;
-
-			match active_head.waiting_large_statements.entry(candidate_hash) {
-				Entry::Occupied(mut occupied) => {
-					match occupied.get_mut() {
-						LargeStatementStatus::Fetching(info) => {
-							if info.available_peers.insert(peer, metadata).is_none() {
-								info.peers_to_try.push(peer);
-								// Answer any pending request for more peers:
-								if let Some(sender) = std::mem::take(&mut info.peer_sender) {
-									let to_send = std::mem::take(&mut info.peers_to_try);
-									if let Err(peers) = sender.send(to_send) {
-										// Requester no longer interested for now, might want them
-										// later:
-										info.peers_to_try = peers;
-									}
-								}
+	match active_head.waiting_large_statements.entry(candidate_hash) {
+		Entry::Occupied(mut occupied) => {
+			match occupied.get_mut() {
+				LargeStatementStatus::Fetching(info) => {
+
+					let is_new_peer = !info.available_peers.contains_key(&peer);
+					let is_large_statement = message.is_large_statement();
+
+					match info.available_peers.entry(peer) {
+						Entry::Occupied(mut occupied) => {
+							occupied.get_mut().push(message);
+						}
+						Entry::Vacant(vacant) => {
+							vacant.insert(vec![message]);
+						}
+					}
+
+					if is_new_peer & is_large_statement {
+						info.peers_to_try.push(peer);
+						// Answer any pending request for more peers:
+						if let Some(sender) = std::mem::take(&mut info.peer_sender) {
+							let to_send = std::mem::take(&mut info.peers_to_try);
+							if let Err(peers) = sender.send(to_send) {
+								// Requester no longer interested for now, might want them
+								// later:
+								info.peers_to_try = peers;
```

Comment:
I assume we don't restore the `info.sender` because it has been consumed and will be set again later.

Reply:
Yes, requester will issue new request when it feels a desire for more peers again.